### PR TITLE
Donated item list

### LINKED ISSUE #40: Replace fake Donated Items with real Donations
Replace the hardcoded Donated Items list currently displayed in a tabular format with real, dynamic data. This will utilize data from the newly implemented Donated Item Form. Additionally, remove the unnecessary Barcode field and ensure all other fields remain visible. Update the donor field to show donor names (instead of just email) once the corresponding API is ready, and maintain existing filter and sorting features on the dynamic data.

**Current Situation:**

- The list of donations is currently displayed using hardcoded data, which limits its usefulness and relevance.
- A new Donated Item Form is now available that can be used to add new donations, making it possible to replace the hardcoded data with real dynamic data fetched from the backend.
- The Barcode field is no longer necessary, and we need to remove it from the table display.
- The current Donor Email field only shows an email string. Once the API is ready to provide all donor details, the Donor Name should be displayed instead of just the email.

Requirements:

- [x] **Replace Fake Data with Dynamic Data:**

- Populate the Donated Items table with real dynamic data fetched from the backend.
- Remove hardcoded data and replace it with real donations obtained via the appropriate API.

- [x] **Remove Barcode Field:**

- Remove the Barcode field from the displayed table, as it is no longer relevant.
- Ensure the remaining fields are still present and correctly displayed, including Item ID, Item Type, Current Status, Date Donated, Donor Email (or Donor Name), and Associated Program.
 

- [x] **Update Donor Field:**

For now, keep displaying the Donor Email in the table.
Once the API is updated to provide detailed donor information, update the table to show the Donor Name instead of just the email.
Ensure the frontend is set up to dynamically handle the updated data without requiring further changes to the logic.

- [x] **Ensure Filters and Sorting Work:**

Ensure the existing filters (e.g., by status, item type, program) continue to work seamlessly with the dynamic data.
Ensure the sorting functionality (e.g., sorting by date, donor, item type) still works as expected with the real donation data.

**Write Unit Tests:**
Write unit tests to verify that:

- The Donated Items table is correctly populated with real data from the backend.
- The Donor Name field is displayed correctly once donor details are provided via the updated API.
- The filter and sorting functionalities are functioning as expected with the new dynamic data.

Tests should cover scenarios such as:

- Successful Data Fetching: Ensure the data is successfully fetched from the backend and displayed in the table.
- Edge Cases: Verify that the table works correctly with empty data, partial data, and unexpected inputs.
- Filter and Sorting: Confirm that filtering and sorting features return accurate results based on the dynamic data.

### PR DESCRIPTION
**Pull request by miabu-pashh**
# Pull Request Template

**Fixes #40**

### What was changed?

Replaced the hardcoded Donated Items list currently displayed in a tabular format with real, dynamic data. This will utilize data from the newly implemented Donated Item Form. Maintained the existing filter and sorting features on the dynamic data.

**Example:**

- Updated the donated item list dynamically instead of hardcoded values.

### Why was it changed?

- donated item list was showing the hardcoded values of the donated items, it was necessary to show the real data and also the barcode for them to view and downlaod a barcode.

### How was it changed?

-Creating a type scriipt file and enhancing the ui part by creating a separate file for css .

### Screenshots that show the changes (if applicable):

- **Before:**
  - 
- **After:**
image

### CODE REVIEW

File: client-app/tsconfig.json [hunk 1]
```diff
@@ -1,20 +1,37 @@
...
         /* Visit https://aka.ms/tsconfig to read more about this file */
 
         /* Language and Environment */
-        "target": "es2016" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
+        // "target": "es2016" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
 
-        /* Modules */
-        "module": "commonjs" /* Specify what module code is generated. */,
-        "esModuleInterop": true /* Emit additional JavaScript to ease support for importing CommonJS modules. This enables 'allowSyntheticDefaultImports' for type compatibility. */,
-        // "preserveSymlinks": true,                         /* Disable resolving symlinks to their realpath. This correlates to the same flag in node. */
```
Comment (saikiran0405) [inline client-app/tsconfig.json:11]: It's good to remove unnecessary comments

File: server/src/__tests__/donatedItem.test.ts [hunk 2]
```diff
@@ -16,38 +16,48 @@ describe('DonatedItem API Tests', () => {
...
+            ...updateItem,
+        });
     });
 
     // Test POST /donatedItem
     it('validates Program name and Donor email before creating a donated item', async () => {
         const response = await request(app).post('/donatedItem').send(newItem);
         expect(response.status).toBe(201);
         expect(mockPrismaClient.donatedItem.create).toHaveBeenCalled();
-
```
Comment (saikiran0405) [inline server/src/__tests__/donatedItem.test.ts:28]: Consider creating a helper function for validation to avoid code duplication in POST and PUT routes.

File: server/src/__tests__/donatedItem.test.ts [hunk 3]
```diff
@@ -16,38 +16,48 @@ describe('DonatedItem API Tests', () => {
...
+        expect(response.body.error).toContain(
+            `Donor with ID: ${donorId} does not exist.`,
+        );
     });
 
     // Test PUT /donatedItem/details/{id}
     it('updates donated item details correctly', async () => {
-        const response = await request(app).put('/donatedItem/details/1').send(updateItem);
+        const response = await request(app)
+            .put('/donatedItem/details/1')
```
Comment (saikiran0405) [inline server/src/__tests__/donatedItem.test.ts:59]: Ensure that all fields updated in the PUT request are properly verified in the test.

File: client-app/src/css/DonatedItemsList.css [hunk 4]
```diff
@@ -1,13 +1,15 @@
 .item-list {
-    width: 100%;
+    width: 90%;
     border-collapse: collapse;
+    margin: 0 auto; /* This will center the table horizontally */
 }
 
 .item-list th,
```
Comment (saikiran0405) [inline client-app/src/css/DonatedItemsList.css:7]: It seems we can improve a little this css file .

File: client-app/src/Components/DonatedItemsList.tsx [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+    const [loading, setLoading] = useState<boolean>(true);
+    const [error, setError] = useState<string | null>(null);
+    const [programOptions, setProgramOptions] = useState<Program[]>([]);
+    const [selectedProgram, setSelectedProgram] = useState<string>('');
+    const [assignProgramClicked, setAssignProgramClicked] =
+        useState<boolean>(false);
+    const [modalIsOpen, setModalIsOpen] = useState<boolean>(false);
+    const [itemTypes, setItemTypes] = useState<Set<string>>(new Set());
+
+    const navigate = useNavigate();
```
Comment (saikiran0405) [inline client-app/src/Components/DonatedItemsList.tsx:61]: Ensure all relevant state and props that affect this useEffect are included in the dependency array to prevent stale state issues.

File: client-app/src/Components/DonatedItemsList.tsx [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+
+    const fetchDonatedItems = async (): Promise<void> => {
+        try {
+            setLoading(true);
+            const response = await fetch(
+                `${process.env.REACT_APP_BACKEND_API_BASE_URL}/donatedItem`,
```
Comment (saikiran0405) [inline client-app/src/Components/DonatedItemsList.tsx:67]: Consider wrapping this and similar functions with useCallback to prevent unnecessary re-renders, especially if these handlers are passed as props to child components.

File: client-app/src/Components/DonatedItemsList.tsx [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+            item =>
+                item.currentStatus.toLowerCase() ===
+                event.target.value.toLowerCase(),
+        );
+        setFilteredItems(filtered);
+    };
+
+    const toggleAssignProgram = (): void => {
+        setAssignProgramClicked(!assignProgramClicked);
+    };
```
Comment (saikiran0405) [inline client-app/src/Components/DonatedItemsList.tsx:243]: Add error handling for the barcode download process to manage failures in html2canvas or any other issues

File: client-app/src/Components/DonatedItemsList.tsx [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+        <div>
+            <div className="header">
+                <div className="logo-container">
+                    <img
+                        src="https://www.bworks.org/wp-content/themes/bworks/library/images/logo-bworks.png"
+                        alt="BWorks Logo"
+                        className="logo"
+                    />
+                </div>
+                <div className="options">
```
Comment (saikiran0405) [inline client-app/src/Components/DonatedItemsList.tsx:287]: Consider extracting this button into a separate reusable component to reduce code duplication and improve maintainability.

**Review by saikiran0405 — COMMENTED** [review]
1. Remove Unnecessary code or comments
2. Revert some files , it's heavy to review these many files
3. Include all relevant state and props in the useEffect dependency array to prevent stale data issues.
4. Wrap functions with `useCallback` to prevent unnecessary re-renders, especially when passing handlers to child components.
5. Add error handling for the barcode download process
6. Extract the button into a reusable component to reduce duplication and enhance maintainability.
7. Only one option is available in the drop-down in forntend
8. Need to work more on DonateItemList.css

File: client-app/tsconfig.json [hunk 9]
```diff
@@ -1,20 +1,37 @@
...
-        /* Type Checking */
-        "strict": true /* Enable all strict type-checking options. */,
-        "skipLibCheck": true /* Skip type checking all .d.ts files. */,
-        "jsx": "react-jsx"
+        // /* Type Checking */
+        // "strict": true /* Enable all strict type-checking options. */,
+        // "skipLibCheck": true /* Skip type checking all .d.ts files. */,
+        // "jsx": "react-jsx",
+
+        "target": "es5",
```
Comment (Anjali0407-git) [inline client-app/tsconfig.json:20]: Did you have any plan to change target from es2016 to es5? Is this all new configs necessary?

File: client-app/src/Components/DonatedItemsList.tsx (hunk 6 above)
Comment (Anjali0407-git) [reply client-app/src/Components/DonatedItemsList.tsx:67]: Remove the extra `'/' ` before donatedItem in the above URL

File: client-app/src/Components/DonatedItemsList.tsx [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+    lastUpdated?: string;
+    donor: Donor | null;
+    program: Program | null;
+    programId?: number;
+    statuses?: Status[];
+}
+
+interface SelectedItemDetails extends DonatedItem {
+    statuses: Status[];
+}
```
Comment (Anjali0407-git) [inline client-app/src/Components/DonatedItemsList.tsx:42]: Can you move them into modals folder so that they will be reusable

**Review by Anjali0407-git — CHANGES_REQUESTED** [review]

File: client-app/tsconfig.json (hunk 9 above)
Comment (miabu-pashh) [reply client-app/tsconfig.json:20]: I found this advantages: Strict mode, which allows developers to opt into a restricted variant of JavaScript 
JSON support, which streamlines data interchange in web applications 
New methods for arrays and strings, including forEach, map, filter, and reduce

File: package.json [hunk 11]
```diff
@@ -1,6 +1,13 @@
 {
   "devDependencies": {
-    "prettier": "^3.3.3"
+    "@types/html2canvas": "^1.0.0",
+    "@types/react": "^18.3.12",
```
[the remark this answers is not in the document]
Comment (miabu-pashh) [reply package.json:4]: we can start the app by in core directory both backend and front end., so. I guess this would help.

**Comment by miabu-pashh** [comment]
I'm doing  npm run prettier:write, and NPM run prettier:check every time I push to main, so its changing the format, lines and spaces. i guess thats the reason many files are affecting , I mean their format is changing to good.

**Comment by Anjali0407-git** [comment]
@miabu-pashh is this ready for re-review?

File: package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (Anjali0407-git) [reply package.json:4]: @miabu-pashh We do not have any react code in root. It's in /client-app directory. Please revert all changes in this file

File: client-app/src/Components/Programs.tsx [hunk 12]
```diff
@@ -20,7 +20,7 @@ function Programs() {
         const fetchPrograms = async () => {
             try {
                 const response = await axios.get<Program[]>(
-                    `${process.env.REACT_APP_BACKEND_API_BASE_URL}program`,
+                    `${process.env.REACT_APP_BACKEND_API_BASE_URL}/program`,
```
Comment (Anjali0407-git) [inline client-app/src/Components/Programs.tsx:23]: @miabu-pashh Why did you bring up this change. I already mentioned to follow the existing convention and remove the `'/'` before `program`.

Please update the backend api variable in .env file to `REACT_APP_BACKEND_API_BASE_URL="http://localhost:4010/"`
There is already a leading '/' in this file, so you do not have to put it here.

File: server/package.json [hunk 13]
```diff
@@ -4,7 +4,7 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "start": "node dist/index.js",
+    "start": "concurrently \"react-scripts start\" \"cd server && nodemon server\"",
```
Comment (Anjali0407-git) [inline server/package.json:7]: @miabu-pashh What command did you use to start up the server? This change is not related to this issue. If you want to update any configs or startup scripts, create an another PR for that. DO NOT include all of them in a single PR. Please revert this file. Please only commit the relevant files to your branch.

File: client-app/tsconfig.json (hunk 9 above)
Comment (Anjali0407-git) [reply client-app/tsconfig.json:4]: @miabu-pashh strict mode was already true. I would suggest to revert this config file entirely. Please create a new PR for such updates in future
